api/v1alpha2/loadbalancer/types.go

```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package loadbalancer
+
+import (
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// AwsProvider is the name of the aws provider
+	AwsProvider = "aws"
+)
+
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=awsloadbalancers,scope=Namespaced,categories=cluster-api
+// +kubebuilder:storageversion
+// +kubebuilder:subresource:status
+
+// AWSLoadBalancer is the schema for the AWS Load balancer API
+type AWSLoadBalancer struct {
```

Comment:
Shouldn't the API types be generic and the implementations be provider specific?

Reply:
No. Think of this as a KubeadmBootstrap provider. There will be a diff controller / API model for each implementation of the LB.

Reply:
The issue is that there's no common API for loadbalancers, as each provider needs its own information. I don't want us to endup implementing annotations on-top to satisfy each provider's specifics.

Another way could be store the information on-disk/cm implement something `Service` (e.g. `MachineService` or `MachineLoadbalancer`) which would select machines based on a the selector and add them to the load balancing pool. A drawback is that a management cluster is restricted to provision against one Loadbalancing provider. Thoughts @akutz @andrewsykim ?

Reply:
I don't think it's that complicated. Just like each `Cluster` resource today has a `ConfigRef` to a Infrastructure provider's `Cluster`, so to will there be a `ConfigRef` from a `Cluster` to a load balancer config, in this case, `AWSLoadBalancerConfig`.

There's some questions around how to get the IP information from the machines, but that's something we can work out. 

The benefit of this decoupled model is that it enables the introduction of a load balancer provider of any kind.

Reply:
Yes, plus this wouldn't change much the implementation itself. thoughts @andrewsykim ?

Reply:
So my thoughts around this was that we introduce a generic `LoadBalancer` type with a `ProviderRef` field similar to what @akutz mentioned re: Cluster with a ConfigRef. That way, for each implementation you only have to implement a new ProviderRef and not the entire LoadBalancer API again. Maybe we already agree on this but it's not clear to me yet.